packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -252,40 +275,36 @@ export class TokenDetectionController extends PollingControllerV1<
   async detectTokens(options?: {
     networkClientId?: NetworkClientId;
     accountAddress?: string;
-  }) {
-    const { networkClientId, accountAddress } = options || {};
-    const {
-      disabled,
-      isDetectionEnabledForNetwork,
-      isDetectionEnabledFromPreferences,
-    } = this.config;
+  }): Promise<void> {
+    const { networkClientId, accountAddress } = options ?? {};
     if (
-      disabled ||
-      !isDetectionEnabledForNetwork ||
-      !isDetectionEnabledFromPreferences
+      this.#disabled ||
+      !this.#isDetectionEnabledForNetwork ||
+      !this.#isDetectionEnabledFromPreferences
     ) {
       return;
     }
-    const { tokens } = this.getTokensState();
-    const selectedAddress = accountAddress || this.config.selectedAddress;
-    const chainId = this.getCorrectChainId(networkClientId);
-
-    const tokensAddresses = tokens.map(
+    const { tokens } = this.#getTokensState() ?? {};
```

Comment:
Do we need the `?? {}` here?

Reply:
Removed: 720dc698e6faa4d8a5f4c07cf3996453f03d7f26

We do get the following runtime error here if the `getTokensState` callback doesn't return a valid tokens state object e.g. if `new TokensDetectController({ ..., getTokensState: sinon.stub() })`:
```
 TypeError: Cannot destructure property 'tokens' of '((cov_1sp0cd8ows(...).s[79]++) , __classPrivateFieldGet(...).call(...))' as it is undefined.
```

We could rely on type checking to enforce the return type of `getTokensState`, but using `?? {}` and `tokens?.` will ensure that `detectTokens()` never fails at runtime.

Reply:
Ah, I see. It looks like `sinon.stub()` has a type of `sinon.SinonStub<any, any>` (and `jest.fn()` has a type of `jest.Mock<any, any>`), so we should have gotten a type error but did not. That seems to be a problem with using mock functions to fill in arguments in general: if you don't supply a return value, it may cause the code you're testing to break in some way. I suspect that this pattern is used in all kinds of tests, so if we really wanted to handle this, we'd have to add runtime checks to controllers that take functions in this way. But... that doesn't feel great.

It seems that there are a couple of long-term fixes we could make which would make this problem go away:

1. We have many, many tests that use Sinon's mock functions when we should be using Jest's instead. This is a long-standing need, and this reminds that I should make tickets to capture the work to convert that syntax.
2. I just checked and apparently Jest now ships with TypeScript types that we're not using (we're using `@types/jest` instead, which is supported by the community). And in Jest's version of the types, the type of `jest.fn()` returns `jest.Mock<UnknownFunction>`. So, after we upgrade to Jest 29, perhaps we can look into dropping `@types/jest` in favor of Jest's types? It would probably require some substantial changes due to how Jest recommends importing types now: https://jestjs.io/docs/getting-started#type-definitions

What are your thoughts? Should we use a short-term fix now or wait for the long-term fix?

Reply:
Not adding runtime checks feels a little iffy, but I agree that this shypesould definitely be caught at the type level. The `any` typing suppressing these errors seems like a pretty critical issue. I ran into it multiple times fixing tests here. I like the long-term option of transitioning to `jest.fn()` and built-in jest types. I'll keep digging into the jest update PR I have open: https://github.com/MetaMask/core/pull/3646.

Reply:
Related to this, `ReturnType<NetworkController['getNetworkClientById']>` always defaulting to `AutoManagedNetworkClient<CustomNetworkClientConfiguration>` is a headache, as it leads to `jest.fn()<ReturnType<...>, Parameters<...>>` and `sinon.stub<Parameters<...>, ReturnType<...>>()` not accepting infura networks. Perhaps it might be worth exploring a different strategy than method overloading for `getNetworkClientById` (e.g. parametric polymorphism, composition, dependency injection)?